File: Depth-Perception-Using-Stereo-Camera/cpp/disparity2depth_calib.cpp
#include <cmath>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

#include "stereo_depth.hpp"

namespace {

constexpr const char* kWindow = "Depth calibration disparity";

struct Options {
    int leftCamera = 2;
    int rightCamera = 0;
    double maxDistance = 230.0;
    double minDistance = 50.0;
    double sampleStep = 40.0;
    std::filesystem::path maps =
        std::filesystem::path(STEREO_SOURCE_DIR) / "data" /
        "stereo_rectify_maps.xml";
    std::filesystem::path config =
        std::filesystem::path(STEREO_SOURCE_DIR) / "data" /
        "depth_estimation_params_cpp.xml";
    std::filesystem::path outputConfig =
        std::filesystem::path(STEREO_SOURCE_DIR) / "data" /
        "depth_estimation_params_cpp_updated.xml";
};

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        auto requireValue = [&](const char* option) -> std::string {
            if (index + 1 >= argc) {
                throw std::invalid_argument(
                    std::string("Missing value for ") + option);
            }
            return argv[++index];
        };
        if (argument == "--left-camera") {
            options.leftCamera =
                std::stoi(requireValue("--left-camera"));
        } else if (argument == "--right-camera") {
            options.rightCamera =
                std::stoi(requireValue("--right-camera"));
        } else if (argument == "--maps") {
            options.maps = requireValue("--maps");
        } else if (argument == "--config") {
            options.config = requireValue("--config");
        } else if (argument == "--output-config") {
            options.outputConfig = requireValue("--output-config");
        } else if (argument == "--max-distance") {
            options.maxDistance =
                std::stod(requireValue("--max-distance"));
        } else if (argument == "--min-distance") {
            options.minDistance =
                std::stod(requireValue("--min-distance"));
        } else if (argument == "--sample-step") {
            options.sampleStep =
                std::stod(requireValue("--sample-step"));
        } else {
            throw std::invalid_argument("Unknown option: " + argument);
        }
    }
    if (options.leftCamera == options.rightCamera) {
        throw std::invalid_argument(
            "Left and right camera indices must be different.");
    }
    if (
        options.minDistance <= 0.0 ||
        options.maxDistance <= options.minDistance ||
        options.sampleStep <= 0.0) {
        throw std::invalid_argument(
            "Distance limits and sample step are invalid.");
    }
    return options;
}

struct CalibrationState {
    cv::Mat disparity;
    double targetDepth = 0.0;
    double sampleStep = 0.0;
    int minDisparity = 0;
    std::vector<double> disparities;
    std::vector<double> depths;
};

void onMouse(int event, int x, int y, int, void* userData) {
    if (event != cv::EVENT_LBUTTONDOWN || userData == nullptr) {
        return;
    }
    auto& state = *static_cast<CalibrationState*>(userData);
    if (
        state.disparity.empty() || x < 0 || y < 0 ||
        x >= state.disparity.cols || y >= state.disparity.rows) {
        return;
    }
    const float value = state.disparity.at<float>(y, x);
    if (
        !std::isfinite(value) ||
        value <= static_cast<float>(state.minDisparity)) {
        std::cout << "Ignored invalid disparity sample.\n";
        return;
    }
    state.disparities.push_back(value);
    state.depths.push_back(state.targetDepth);
    std::cout << "Depth " << state.targetDepth
              << " cm -> disparity " << value << " px\n";
    state.targetDepth -= state.sampleStep;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const Options options = parseOptions(argc, argv);
        auto config = learnopencv::stereo::loadConfig(options.config);
        const auto maps =
            learnopencv::stereo::loadRectificationMaps(options.maps);
        const auto matcher =
            learnopencv::stereo::createMatcher(config);

        cv::VideoCapture leftCamera(options.leftCamera);
        cv::VideoCapture rightCamera(options.rightCamera);
        if (!leftCamera.isOpened()) {
            throw std::runtime_error(
                "Could not open left camera " +
                std::to_string(options.leftCamera));
        }
        if (!rightCamera.isOpened()) {
            throw std::runtime_error(
                "Could not open right camera " +
                std::to_string(options.rightCamera));
        }

        CalibrationState state;
        state.targetDepth = options.maxDistance;
        state.sampleStep = options.sampleStep;
        state.minDisparity = config.minDisparity;
        cv::namedWindow(kWindow, cv::WINDOW_NORMAL);
        cv::resizeWindow(kWindow, 800, 600);
        cv::setMouseCallback(kWindow, onMouse, &state);

        while (state.targetDepth >= options.minDistance) {
            if (!leftCamera.grab() || !rightCamera.grab()) {
                throw std::runtime_error(
                    "Could not grab synchronized stereo frames.");
            }
            cv::Mat leftColor;
            cv::Mat rightColor;
            if (!leftCamera.retrieve(leftColor) ||
                !rightCamera.retrieve(rightColor) ||
                leftColor.empty() || rightColor.empty()) {
                throw std::runtime_error(
                    "Could not retrieve synchronized stereo frames.");
            }
            cv::Mat leftGray;
            cv::Mat rightGray;
            cv::cvtColor(
                leftColor, leftGray, cv::COLOR_BGR2GRAY);
            cv::cvtColor(
                rightColor, rightGray, cv::COLOR_BGR2GRAY);
            const auto [leftRectified, rightRectified] =
                learnopencv::stereo::rectifyPair(
                    leftGray, rightGray, maps);
            state.disparity =
                learnopencv::stereo::computeDisparity(
                    leftRectified, rightRectified, config, matcher);
            cv::Mat view =
                learnopencv::stereo::disparityVisualization(
                    state.disparity);
            cv::putText(
                view,
                "Place target at " +
                    std::to_string(
                        static_cast<int>(state.targetDepth)) +
                    " cm and click",
                cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 0.65,
                cv::Scalar(255), 2, cv::LINE_AA);
            cv::imshow(kWindow, view);
            if (cv::waitKey(1) == 27) {
                break;
            }
        }
        cv::destroyAllWindows();

        const auto fit = learnopencv::stereo::fitDepthModel(
            state.disparities, state.depths, config.minDisparity);
        config.depthScale = fit.scale;
        config.depthOffset = fit.offset;
        config.validate();
        learnopencv::stereo::saveConfig(
            config, options.outputConfig);
        std::cout << "Depth scale: " << fit.scale << '\n'
                  << "Depth offset: " << fit.offset << '\n'
                  << "Calibration RMSE: " << fit.rmse << " cm\n"
                  << "Saved calibrated configuration: "
                  << options.outputConfig << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}

File: Depth-Perception-Using-Stereo-Camera/python/disparity2depth_calib.py
"""Fit disparity-to-depth scale and offset from a real stereo rig."""

from __future__ import annotations

import argparse
from dataclasses import replace
from pathlib import Path

import cv2
import numpy as np

from stereo_depth import (
    DEFAULT_CONFIG,
    DEFAULT_MAPS,
    PROJECT_DIR,
    compute_disparity,
    create_matcher,
    disparity_visualization,
    fit_depth_model,
    load_config,
    load_rectification_maps,
    rectify_pair,
    save_config,
)


WINDOW = "Depth calibration disparity"
DEFAULT_OUTPUT_CONFIG = (
    PROJECT_DIR / "data" / "depth_estimation_params_py_updated.xml"
)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--left-camera", type=int, default=2)
    parser.add_argument("--right-camera", type=int, default=0)
    parser.add_argument("--maps", type=Path, default=DEFAULT_MAPS)
    parser.add_argument("--config", type=Path, default=DEFAULT_CONFIG)
    parser.add_argument("--output-config", type=Path, default=DEFAULT_OUTPUT_CONFIG)
    parser.add_argument("--max-distance", type=float, default=230.0)
    parser.add_argument("--min-distance", type=float, default=50.0)
    parser.add_argument("--sample-step", type=float, default=40.0)
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if args.left_camera == args.right_camera:
        raise ValueError("Left and right camera indices must be different.")
    if (
        args.min_distance <= 0
        or args.max_distance <= args.min_distance
        or args.sample_step <= 0
    ):
        raise ValueError("Distance limits and sample step are invalid.")

    config = load_config(args.config)
    maps = load_rectification_maps(args.maps)
    matcher = create_matcher(config)
    state: dict[str, object] = {
        "disparity": None,
        "target_depth": float(args.max_distance),
        "disparities": [],
        "depths": [],
    }

    def on_mouse(event: int, x: int, y: int, _flags: int, _param: object) -> None:
        if event != cv2.EVENT_LBUTTONDOWN:
            return
        disparity = state["disparity"]
        if not isinstance(disparity, np.ndarray):
            return
        if not 0 <= y < disparity.shape[0] or not 0 <= x < disparity.shape[1]:
            return
        value = float(disparity[y, x])
        if not np.isfinite(value) or value <= config.min_disparity:
            print("Ignored invalid disparity sample.")
            return
        target_depth = float(state["target_depth"])
        disparities = state["disparities"]
        depths = state["depths"]
        assert isinstance(disparities, list) and isinstance(depths, list)
        disparities.append(value)
        depths.append(target_depth)
        print(f"Depth {target_depth:.2f} cm -> disparity {value:.4f} px")
        state["target_depth"] = target_depth - float(args.sample_step)

    left_camera = cv2.VideoCapture(args.left_camera)
    right_camera = cv2.VideoCapture(args.right_camera)
    if not left_camera.isOpened():
        raise RuntimeError(f"Could not open left camera {args.left_camera}.")
    if not right_camera.isOpened():
        left_camera.release()
        raise RuntimeError(f"Could not open right camera {args.right_camera}.")

    cv2.namedWindow(WINDOW, cv2.WINDOW_NORMAL)
    cv2.resizeWindow(WINDOW, 800, 600)
    cv2.setMouseCallback(WINDOW, on_mouse)
    try:
        while float(state["target_depth"]) >= args.min_distance:
            if not left_camera.grab() or not right_camera.grab():
                raise RuntimeError("Could not grab synchronized stereo frames.")
            left_ok, left_color = left_camera.retrieve()
            right_ok, right_color = right_camera.retrieve()
            if not left_ok or not right_ok:
                raise RuntimeError("Could not retrieve synchronized stereo frames.")

            left_gray = cv2.cvtColor(left_color, cv2.COLOR_BGR2GRAY)
            right_gray = cv2.cvtColor(right_color, cv2.COLOR_BGR2GRAY)
            left_rectified, right_rectified = rectify_pair(
                left_gray, right_gray, maps
            )
            disparity = compute_disparity(
                left_rectified, right_rectified, config, matcher=matcher
            )
            state["disparity"] = disparity
            view = disparity_visualization(disparity)
            cv2.putText(
                view,
                f"Place target at {float(state['target_depth']):.0f} cm and click",
                (10, 30),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.65,
                255,
                2,
                cv2.LINE_AA,
            )
            cv2.imshow(WINDOW, view)
            if cv2.waitKey(1) == 27:
                break
    finally:
        left_camera.release()
        right_camera.release()
        cv2.destroyAllWindows()

    disparities = np.asarray(state["disparities"], dtype=np.float64)
    depths = np.asarray(state["depths"], dtype=np.float64)
    scale, offset, rmse = fit_depth_model(
        disparities, depths, min_disparity=config.min_disparity
    )
    calibrated = replace(
        config, depth_scale=scale, depth_offset=offset
    ).validated()
    destination = save_config(calibrated, args.output_config)
    print(f"Depth scale: {scale:.8f}")
    print(f"Depth offset: {offset:.8f}")
    print(f"Calibration RMSE: {rmse:.6f} cm")
    print(f"Saved calibrated configuration: {destination}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
